podium-gateway/src/main/java/nl/thehyve/podium/service/RequestFileService.java

```diff
@@ -0,0 +1,178 @@
+                                             RequestFileType requestFileType) throws ActionNotAllowed, AccessDenied, IOException {
+
+        Request request = requestRepository.findOneByUuid(requestUuid);
+
+        Boolean allowedToAdd = false;
+
+        Set<UUID> organisationUuids = request.getOrganisations();
+        UUID requester = request.getRequester();
+        UserRepresentation ownerRepresentation = userMapperHelper.uuidToRemoteUserRepresentation(owner.getUserUuid());
+        Set<String> authorities = ownerRepresentation.getAuthorities();
+
+        RequestRepresentation requestRepresentation = requestMapper.overviewRequestToRequestDTO(request);
+        OverviewStatus overviewStatus = requestRepresentation.getStatus();
+
+        // Researchers can add files if it is in Draft(/Revision) status
+        if(authorities.contains(AuthorityConstants.RESEARCHER) &&
+            (overviewStatus == OverviewStatus.Draft || overviewStatus == OverviewStatus.Revision)&&
+            requester.equals(owner.getUserUuid())){
+            allowedToAdd = true;
+        }
+        // Coordinator can add files in Validation status
+        if(securityService.isCurrentUserInAnyOrganisationRole(organisationUuids, AuthorityConstants.ORGANISATION_COORDINATOR)){
+            if(overviewStatus == OverviewStatus.Validation){
+                allowedToAdd = true;
+            }
+        }
+
+        // Reviewer can add files in Review status
+        if(securityService.isCurrentUserInAnyOrganisationRole(organisationUuids, AuthorityConstants.REVIEWER)){
+            if(overviewStatus == OverviewStatus.Review){
+                allowedToAdd = true;
+            }
+        }
+
+        if(!allowedToAdd){
+            throw ActionNotAllowed.forStatus(request.getStatus());
+        }
+
+        RequestFile requestFile = new RequestFile();
+        requestFile.setOwner(owner.getUserUuid());
+
+        requestFile.setRequest(request);
+
+        requestFile.setRequestFileType(requestFileType);
+
+        String uploadDir = podiumProperties.getFiles().getUploadDir();
+        requestFile.setFileName(file.getOriginalFilename());
+        requestFile.setFileByteSize(file.getSize());
+        Path path = Paths.get(uploadDir);
+
+        Path tempFile = Files.createTempFile(path,"", "");
+        Files.write(tempFile, file.getBytes());
+        requestFile.setFileLocation(path.toString());
+
+        requestFileRepository.save(requestFile);
+
+        return requestFileMapper.processingRequestFileToRequestFileDto(requestFile);
+    }
+
+    public ByteArrayResource getFile(IdentifiableUser requester, UUID requestUUID, UUID fileUuid) throws IOException{
+        RequestFile requestFile = requestFileRepository.findOneByUuidAndDeletedFalse(fileUuid);
+
+        Path path = Paths.get(requestFile.getFileLocation());
+        return new ByteArrayResource(Files.readAllBytes(path));
```

Comment:
It looks like this loads the entire file in memory. You could also use, e.g., 
```java
InputStream input = new FileInputStream(path.toFile());
InputStreamResource resource = new InputStreamResource(input);
```
Although, I don't know the implementation details of that.